graspologic/embed/case.py

```diff
@@ -0,0 +1,314 @@
+        graph Laplacian, and :math:`X` is a matrix of covariates for each node.
+
+        Parameters
+        ----------
+        graph : array-like or networkx.Graph
+            Input graph to embed. See graspologic.utils.import_graph
+
+        covariates : array-like, shape (n_vertices, n_covariates)
+            Covariate matrix. Each node of the graph is associated with a set of
+            `d` covariates. Row `i` of the covariates matrix corresponds to node
+            `i`, and the number of columns are the number of covariates.
+
+        y: Ignored
+
+        Returns
+        -------
+        self : object
+            Returns an instance of self.
+        """
+
+        # setup
+        A = import_graph(graph)
+
+        # center and scale covariates to unit norm
+        covariates = normalize(covariates, axis=0)
+        covariates = scale(covariates, axis=0, with_std=False)
+
+        # save necessary params  # TODO: do this without saving potentially huge objects into `self`
+        self._L = _to_reg_laplacian(A)
+        self._R = np.shape(covariates)[1]
+        self._X = covariates.copy()
+
+        # change params based on tuning algorithm
+        if self.embedding_alg == "cca":
+            self._LL = self._L @ self._X
+            self._XXt = 0
+            self.alpha_ = 0
+        elif self.embedding_alg == "assortative":
+            self._LL = self._L
+            self._XXt = self._X @ self._X.T
+            self.alpha_ = self._get_tuning_parameter()
+        elif self.embedding_alg == "non-assortative":
+            self._LL = self._L @ self._L
+            self._XXt = self._X @ self._X.T
+            self.alpha_ = self._get_tuning_parameter()
+
+        self.latent_left_ = _embed(
+            self.alpha_, self._LL, self._XXt, n_clusters=self.n_components
+        )
+        self.is_fitted_ = True
+        # # FOR DEBUGGING  # TODO: remove
+        # kmeans = KMeans(n_clusters=3)
+        # labels_ = kmeans.fit_predict(self.latent_left_)
+        # labels_ = remap_labels(labels, labels_)
+        # print(f"misclustering: {np.count_nonzero(labels - labels_) / len(labels)}")
+
+        # FOR DEBUGGING
+        return self
+
+    def _get_tuning_parameter(self):
+        """
+        Find an alpha within a range which optimizes the k-means objective function on
+        our embedding.
+
+        Parameters
+        ----------
+        LL : array
+            The squared regularized graph Laplacian
+        XXt : array
+            X@X.T, where X is the covariate matrix.
+
+        Returns
+        -------
+        alpha : float
+            Tuning parameter which normalizes LL and XXt.
+        """
+        # TODO: clean this code up if possible
+        # setup
+        if isinstance(self.alpha, (int, float)) and self.alpha != -1:
+            return self.alpha
+        n_clusters = self.n_components  # number of clusters
+        n_cov = self._R  # number of covariates
+        LL = self._LL
+        XXt = self._XXt
```

Comment:
This seems redundant esp cause you have delayed(_cluster)(alpha, LL=self._LL, XXt=self._XXt, n_clusters=n_clusters) on line 262. Is the copy necessary?

Reply:
irrelevant now cause I ripped out the tuning algorithm because it was annoying and usually didn't actually improve anything over just using the leading eigenvalues (and in fact often made the embedding worse)